### PR TITLE
Adding regex support for hashmod relabel action

### PR DESCRIPTION
**Pull request by rbizos**
**What**
Adding support for regex on the hashmod action in relabel config to apply only if matching a certain pattern. hashmod is currently the only relabel action that doesn't support regex, adding this will add consistency across all actions. 

**Why**
Label with two formats from the same target. One of which has a high cardinality.
example: 
- good format: instance_id within boundaries (for instance 2x instance numbers)
- high cardinality format: instance_id being host:port creating potential hundred of thousands of 

We want to limit the cardinality of the second by applying a modulus on it an limit the effects like this:

```
metric_relabel_configs:
    - source_labels: [instance_id]
      action: hashmod
      modulus: 1000
      target_label: '__tmp_id'
    - source_labels: [instance_id,__tmp_id]
      regex: '[a-z]+:\d+;(\d+)'
      target_label: 'instance_id'
      replacement: '$1'
    - action: labeldrop
      regex: '__tmp_id'
```

This could be written to:

```
 metric_relabel_configs:
      - source_labels: [instance_id]
        regex: '[a-z]+:\d+'
        modulus: 1000
        action: hashmod
        target_label: 'instance_id'
```

This adds simplicity and only hash if its needed

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thank you, can you also update the documentation? Also, I think we could just have one test, with 4 inpiut labelsets (2 with hasmods, 2 without, interleaved).

**Comment by rbizos** [comment]
Sure

File: pkg/relabel/relabel_test.go [hunk 1]
```diff
@@ -233,6 +233,54 @@ func TestRelabel(t *testing.T) {
 				"d": "976",
 			}),
 		},
+		{
+			input: labels.FromMap(map[string]string{
+				"a": "foo",
+				"b": "foo-bar",
+				"c": "bar",
+				"d": "baz",
```
Comment (roidelapluie) [inline pkg/relabel/relabel_test.go:241]: Sorry!! I did not pay attention, I was thinkin that break was breaking a loop :) Not I see it is a switch. I am happy to keep only "a" and "b:"

Thanks again!!!

Comment (rbizos) [reply pkg/relabel/relabel_test.go]: No worry at all

**Comment by roidelapluie** [comment]
Thanks. Currently this is can be a breaking change and I need to think about this.

The breaking change is that now, in this situation, with a newline in the label, not all the labels will be hashmaped:

```
scrape_configs:
  - job_name: 'demo'
    static_configs:
      - targets:
        - 127.0.0.1:9090
        labels:
          foo: bar
      - targets:
        - 127.0.0.1:8000
        labels:
          foo: |
            foo
            baz
    relabel_configs:
      - source_labels: [foo]
        target_label: foo
        action: hashmod
        modulus: 400
```

**Comment by roidelapluie** [comment]
Sorry that I did not realize that sooner! This is because the default regex is `.*`,  which does not include newlines.

The options would be:
- Ignore when the regex is `.*` (technically not correct)
- Change the default (would affect all? => breaking change)
- Keep as is, don't merge this pull request
- ... ?

Let's get feedback from @beorn7 @juliusv on this one.

**Comment by roidelapluie** [comment]
I have open #8523 to have an explicit test. In my personal opinion, I would be tempted to close this as there is no non-breaking and technically correct solution for this problem.

I would also note that besides the additional lines, there should be close to none performance impact as we do not compute that every time. Once we have relabeled a target, we do not relabel again, but we remember the result.

**Comment by rbizos** [comment]
Good catch I didn't think about this case. 

Yes indeed performance gain are at best marginal. Something possible is change the default regexp for this action alone like this to prevent any breaking change.
```
	if c.Action == HashMod && c.Regex.String() == "(.*)" {
		c.Regex = MustNewRegexp("(?s)(.*)")
	}
```
I'm not certain it's worth adding a special case for this corner case though

**Comment by beorn7** [comment]
_Very_ generally I think the default regexp (who the f*** called it _regex_?!? – an abomination for every decent Gopher… :) should match _any_ label value always. I'm pretty sure it was an oversight that it would not match label values with a newline. _But_ we can only change that in Prometheus 3.

In the meantime, I think we can special-case that the default regexp is `(?s)(.*)` if the action in `hashmod`. (But it has to be implemented differently than as suggested by @rbizos because it still has to be possible to specify `(.*)` if you actually do _not_ want to match newlines. It will be a bit messy, but I think that's fine because all will be fine again in Prom3 when the default regexp will be `(?s)(.*)`.)

Perhaps, just in case, we should quickly find out if there is consensus to change the default regexp in Prom3 (so that we can confident that the quirk we would add now will un-quirk itself in the future).

**Comment by beorn7** [comment]
(The use case here is probably somewhat weird and borderline questionable. But as @rbizos says, ignoring the regexp just for hashmod, as we do now, is not consistent.)

**Comment by juliusv** [comment]
> We want to limit the cardinality of the second by applying a modulus on it an limit the effects like this:
> 
> ```
> metric_relabel_configs:
>     - source_labels: [instance_id]
>      action: hashmod
>      modulus: 1000
>      target_label: '__tmp_id'
>    - source_labels: [instance_id,__tmp_id]
>      regex: '[a-z]+:\d+;(\d+)'
>      target_label: 'instance_id'
>      replacement: '$1'
>    - action: labeldrop
>      regex: '__tmp_id'
>```

@rbizos I'm not sure I understand this use case correctly. There is no `keep`-sharding happening based on the modulus in those relabeling rules. Instead, the modulus is stored in the `instance_id` label, causing many target identities to be mapped into fewer, which would cause target label / series collisions and thus errors... am I missing something?

Also, I don't think the `labeldrop` step at the end is needed for anything, as labels starting with two underscores are removed after relabeling anyway.

**Comment by beorn7** [comment]
Yes, I think @juliusv is correct here. To really reduce cardinality, you need scrape-time evaluations, which is a feature quite far down in the pipeline.

**Comment by rbizos** [comment]
@juliusv I don't think you missed anything. The use-case is indeed corner-case and somewhat hacky. This aimed to be a temporary fix to allow application to keep exposing the 2 labels formats. Collisions are possible indeed, the idea was having a modulo value sufficiently bigger than the number of parallel instances. 

> Also, I don't think the labeldrop step at the end is needed for anything, as labels starting with two underscores are removed after relabeling anyway.
That's a copy paste error on my side, as I played with those rules. I was using this rules in a write_relabel_config and AFAIK the double underscore drop only works on relabel_configs.

**Comment by roidelapluie** [comment]
We do not remove those labels after metrics relabel configs, only after targets relabeling.

I am happy to change the default regex in Prometheus 3.

I think this pull request is fine, I did not dig too much into the use case itself.

**Comment by roidelapluie** [comment]
I think that because it requires an exception for hash mod, I would be tempted to keep this open and tagged for Prometheus 3.

**Comment by juliusv** [comment]
> Collisions are possible indeed, the idea was having a modulo value sufficiently bigger than the number of parallel instances.

How would that solve the cardinality issue then?

> I would be tempted to keep this open and tagged for Prometheus 3.

:+1:

**Comment by rbizos** [comment]
> How would that solve the cardinality issue then?
It prevents the cardinality to increase over time as the instances restarts. In this case 1000 << (n hosts * 65k ports)

**Comment by juliusv** [comment]
Ah ok, yeah.

**Comment by roidelapluie** [comment]
It would also be possible to considerer the default `(.*)` regex as a "bug" and fix it to be `(?s)(.*)` by default. We would mark it as a change/bugfix in the release notes.

**Comment by roidelapluie** [comment]
I have added https://github.com/prometheus/prometheus/issues/8525 for the regex thing.

I think that because 99.9% of the users would not need this for hashmod, and the for .1% there is an alternative - which does not have performance impact (as you proposed), we could simply put this "in the fridge" and see once #8525 is closed what we do with this pull request.

**Comment by roidelapluie** [comment]
We have looked at this during our bub scrub, and because we are planning to fix the default regex in Prometheus 3 (#8525), we are going to close this pull request now and revisit this after the 3.x release. In the meantime, it seems you have provided an effective workaround, so that should not block you. Thanks!

**Comment by beorn7** [comment]
I have also added a reminder to #8525 about this.

**Comment by rbizos** [comment]
Nothing blocking indeed. Thanks for your time !

**Pull request closed without merging**